services/QuillLMS/app/models/session_feedback_history.rb

```diff
@@ -0,0 +1,87 @@
+# A specific aggregation pattern for FeedbackHistory records centered around
+# `activity_session_uid`
+class SessionFeedbackHistory
+  DEFAULT_PAGE_SIZE = 25
+
+  def self.list_by_activity_session(activity_id: nil, page: 1, page_size: DEFAULT_PAGE_SIZE)
+    query = FeedbackHistory.select('feedback_histories.activity_session_uid AS session_uid, min(feedback_histories.time) as start_date, comprehension_prompts.activity_id, max(because_feedback.attempt) AS because_attempts, max(but_feedback.attempt) AS but_attempts, max(so_feedback.attempt) AS so_attempts, (bool_or(because_feedback.optimal)::integer & bool_or(but_feedback.optimal)::integer & bool_or(so_feedback.optimal)::integer)::boolean AS complete')
+      .joins("LEFT OUTER JOIN comprehension_prompts ON feedback_histories.prompt_id = comprehension_prompts.id")
+      .joins("LEFT OUTER JOIN feedback_histories AS because_feedback ON feedback_histories.id = because_feedback.id AND comprehension_prompts.conjunction = 'because'")
+      .joins("LEFT OUTER JOIN feedback_histories AS but_feedback ON feedback_histories.id = but_feedback.id AND comprehension_prompts.conjunction = 'but'")
+      .joins("LEFT OUTER JOIN feedback_histories AS so_feedback ON feedback_histories.id = so_feedback.id AND comprehension_prompts.conjunction = 'so'")
+      .where(used: true)
+      .group(:activity_session_uid, :activity_id)
+      .order('start_date DESC')
+    query = query.where(comprehension_prompts: {activity_id: activity_id.to_i}) if activity_id
+    query = query.limit(page_size)
+    query = query.offset((page - 1) * page_size.to_i) if page && page > 1
+    query
+  end
+
+  def self.serialize_list_by_activity_session(results)
+    results.map do |r|
+      {
+        session_uid: r.session_uid,
+        start_date: r.start_date.iso8601,
+        activity_id: r.activity_id,
+        because_attempts: r.because_attempts || 0,
+        but_attempts: r.but_attempts || 0,
+        so_attempts: r.so_attempts || 0,
+        complete: r.complete || false
+      }
+    end
+  end
+
+  def self.serialize_detail_by_activity_session(activity_session_uid)
```

Comment:
For the counts and aggregage numbers, I think it would be better to reuse the aggregate query above and just add a `where(activity_session_uid: activity_session_uid)`. It feels weird to use different calculation methods for the same numbers.

I think this method should be revisited, happy to jump on a call to talk through some options.

Reply:
So, it may not be clear, but the payload shapes for the "index" or "list" view above and the "data for a single session" here is actually significantly different, which is why I didn't just limit the scope of the list query for the details version.  There's a lot more data in the details query.

https://www.notion.so/quill/Feedback-History-API-Spec-464be720cf034f429c5f82a75776e65c